src/distances.jl

```diff
@@ -19,12 +19,44 @@ from the unprimed to the primed indices (if they are matrix-like).
 
 Matrix-like ITensors should be Hermitian and non-negative.
 """
-function fidelity(ρ::ITensor, σ::ITensor)
-  @assert order(ρ) == order(σ)
-  ρ ./= tr(ρ)
-  σ ./= tr(σ)
-  F = product(product(sqrt_hermitian(ρ), σ), sqrt_hermitian(ρ))
-  F = real(tr(sqrt_hermitian(F)))^2
+function fidelity(A::ITensor, B::ITensor; process::Bool = false)
```

Comment:
Function to handle the fidelity of two input ITensors, without type information to help choosing the correct metric. Requires lots of checking tags.

Reply:
We need to think about a better code design for avoiding this kind of complexity.

I'm imagining we could have an internal wrapper type, something like `ITensorOperator` and `ITensorState` defined as:
```julia
struct ITensorOperator
  t::ITensor
end

struct ITensorState
  t::ITensor
end
```
An `ITensorOperator` represents that an `ITensor` has pairs of primed and unprimed indices which can be treated as a matrix, while an `ITensorState` represents that it just has unprimed indices and is treated as a vector. Then we can have a function like:
```julia
operator_or_state(A::ITensor) = is_operator(A) ? ITensorOperator(A) : ITensorState(A)

fidelity(A::ITensor, B::ITensor) = fidelity(operator_or_state(A), operator_or_state(B))

fidelity(A::ITensorOperator, B::ITensorOperator) = [...]
fidelity(A::ITensorOperator, B::ITensorState) = [...]
# etc.
```
We would need to incorporate the Choi parts as well, I would need a reminder about all of the cases this function is supposed to handle. But the basic idea would be to define temporary wrapper types so we can factorize the function into different pieces using types.

Reply:
Yeah this is great.

Reply:
I am afraid that the issue about MPO density matrix vs MPO unitary remains (i.e. it is not possible to distinguish between them). So the process fidelity between two unitaries, and the state fidelity between two density matrices would have identical inputs.  I think that is why we had to add the `process = true` flag. I'm unsure whether this can be resolved.